service/history/ndc/history_replicator.go

```diff
@@ -227,6 +228,196 @@ func (r *HistoryReplicatorImpl) ApplyEvents(
 	return r.doApplyEvents(ctx, task)
 }
 
+func (r *HistoryReplicatorImpl) BackfillHistoryEvents(
+	ctx context.Context,
+	request *shard.BackfillHistoryEventsRequest,
+) error {
+	task, err := newReplicationTaskFromBatch(
+		r.clusterMetadata,
+		r.logger,
+		request.WorkflowKey,
+		request.BaseExecutionInfo,
+		request.VersionHistoryItems,
+		request.Events,
+		request.NewEvents,
+		request.NewRunID,
+		request.VersionedHistory,
+	)
+	if err != nil {
+		return err
+	}
+
+	return r.doApplyBackfillEvents(ctx, task, r.applyBackfillEvents)
+}
+
+func (r *HistoryReplicatorImpl) doApplyBackfillEvents(
+	ctx context.Context,
+	task replicationTask,
+	action func(context.Context, workflow.MutableState, workflow.Context, wcache.ReleaseCacheFunc, replicationTask) error,
+) (retError error) {
+	wfContext, releaseFn, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		task.getNamespaceID(),
+		task.getExecution(),
+		locks.PriorityHigh,
```

Comment:
Do we need high priority lock for backfill history? Or it does not matter because it should not be the current run of workflow?

Reply:
updated to low priority.